aws/resource_aws_ssm_document.go

```diff
@@ -43,26 +43,30 @@ func resourceAwsSsmDocument() *schema.Resource {
 			"attachments_source": {
 				Type:     schema.TypeList,
 				Optional: true,
+				MaxItems: 20,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"key": {
-							Type:     schema.TypeString,
-							Required: true,
-							ValidateFunc: validation.StringInSlice([]string{
-								ssm.AttachmentsSourceKeyAttachmentReference,
-								ssm.AttachmentsSourceKeySourceUrl,
-								ssm.AttachmentsSourceKeyS3fileUrl,
-							}, false),
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringInSlice(ssm.AttachmentsSourceKey_Values(), false),
 						},
 						"name": {
 							Type:     schema.TypeString,
 							Optional: true,
+							ValidateFunc: validation.All(
+								validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9_\-.]{3,128}$`), ""),
```

Comment:
Reference: https://github.com/hashicorp/terraform-provider-aws/issues/16628

```suggestion
								validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9_\-.]{3,128}$`), "must contain only alphanumeric, underscore, hyphen, or period characters"),
```